lib/galaxy/webapps/galaxy/services/invocations.py

```diff
@@ -0,0 +1,108 @@
+from typing import (
+    Any,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+)
+
+from pydantic import (
+    BaseModel,
+    Field,
+)
+
+from galaxy.exceptions import AdminRequiredException
+from galaxy.managers.histories import HistoryManager
+from galaxy.managers.workflows import WorkflowsManager
+from galaxy.schema.schema import InvocationIndexQueryPayload
+from galaxy.security.idencoding import IdEncodingHelper
+from galaxy.webapps.galaxy.services.base import ServiceBase
+
+
+class InvocationSerializationParams(BaseModel):
+    """Contains common parameters for customizing model serialization."""
+
+    view: Optional[str] = Field(
+        default=None,
+        title="View",
+        description=(
+            "The name of the view used to serialize this item. "
+            "This will return a predefined set of attributes of the item."
+        ),
+        example="element",
+    )
+    step_details: bool = Field(
+        default=False, title="Include step details", description="Include details for individual invocation steps."
+    )
+    legacy_job_state: bool = Field(
+        default=False,
+        deprecated=True,
+    )
+
+
+class InvocationIndexPayload(InvocationIndexQueryPayload):
+    instance: bool = Field(default=False, description="Is provided workflow id for Workflow instead of StoredWorkflow?")
+
+
+class InvocationsService(ServiceBase):
+    def __init__(
+        self, security: IdEncodingHelper, histories_manager: HistoryManager, workflows_manager: WorkflowsManager
+    ):
+        super().__init__(security=security)
+        self._histories_manager = histories_manager
+        self._workflows_manager = workflows_manager
+
+    def index(
+        self, trans, invocation_payload: InvocationIndexPayload, serialization_params: InvocationSerializationParams
+    ) -> Tuple[List[Dict[str, Any]], int]:
+        workflow_id = invocation_payload.workflow_id
+        if invocation_payload.instance:
+            instance = invocation_payload.instance
+            invocation_payload.workflow_id = self._workflows_manager.get_stored_workflow(
```

Comment:
The workflow_id is already decoded at this point (because all of the ids in `InvocationIndexQueryPayload` have the `DecodedDatabaseIdField` type). Ideally we'd change `get_stored_workflow` to work with decoded ids. (I did that in https://github.com/galaxyproject/galaxy/pull/13598/files#diff-15163d5082940fb76cb250da3a76c0efe95acf13ca1c396596c8783c99c23333R94)

Reply:
There was a failing test for this in this branch but it wasn't broken downstream where I've converted more of this to FastAPI I think. Odd. I've "fixed" this by overriding ``galaxy.managers.base.decode_id`` to just convert ``DecodedDatabaseIdField`` to simple integers. I hope this doesn't have any unintended consequences. If it doesn't - perhaps the better approach would be to convert decode_id to always produce a ``DecodedDatabaseIdField`` but that is a bigger change I think.